### PR TITLE
write out prediction json at test step during training

### PR DESCRIPTION
**Pull request by cameron-a-johnson**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **test your PR locally** with `pytest` command?
- [ ] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tcn_hpl/callbacks/plot_metrics.py [hunk 1]
```diff
@@ -362,6 +364,40 @@ def on_test_epoch_end(
...
 
+        # Create activity predictions KWCOCO JSON
+        truth_dset_fpath = trainer.datamodule.hparams["coco_test_activities"]
+        truth_dset = kwcoco.CocoDataset(truth_dset_fpath)
+        acts_dset = kwcoco.CocoDataset()
+        acts_dset.fpath = self._preds_dset_output_fpath
+        acts_dset.dataset['videos'] = truth_dset.dataset['videos']
+        acts_dset.dataset['images'] = truth_dset.dataset['images']
+        acts_dset.index.build(acts_dset)
+        acts_dset.dataset['categories'] = truth_dset.dataset['categories']
```
Comment (Purg) [inline tcn_hpl/callbacks/plot_metrics.py]: I'd still swap the order of these.
```suggestion
        acts_dset.dataset['categories'] = truth_dset.dataset['categories']
        acts_dset.index.build(acts_dset)
```

File: tcn_hpl/callbacks/plot_metrics.py [hunk 2]
```diff
@@ -362,6 +364,40 @@ def on_test_epoch_end(
...
+            video_id = all_source_vids[i].item()
+            # Now get the image_id that matches the frame_index and video_id.
+            sorted_img_ids_for_one_video = acts_dset.index.vidid_to_gids[int(video_id)]
+            image_id = sorted_img_ids_for_one_video[frame_index]
+            # Sanity check: this image_id corresponds to the frame_index and video_id
+            try:
+                assert acts_dset.index.imgs[image_id]['frame_index'] == frame_index
+                assert acts_dset.index.imgs[image_id]['video_id'] == video_id
+            except:
+                import ipdb; ipdb.set_trace()
```
Comment (Purg) [inline tcn_hpl/callbacks/plot_metrics.py]: Take out the debugging
```suggestion
            assert acts_dset.index.imgs[image_id]['frame_index'] == frame_index
            assert acts_dset.index.imgs[image_id]['video_id'] == video_id
```

File: tcn_hpl/callbacks/plot_metrics.py [hunk 3]
```diff
@@ -362,6 +364,40 @@ def on_test_epoch_end(
...
+
+            ann = {
+                "image_id": image_id,
+                "category_id": all_preds[i].item(),
+                "score": all_probs[i][all_preds[i]].item(),
+                "prob": all_probs[i].numpy().tolist(),
+            }
+            acts_dset.add_annotation(**ann)
```
Comment (Purg) [inline tcn_hpl/callbacks/plot_metrics.py]: Creating an extra dictionary structure and then expanding it for the quantity of iterations this loop is likely to run through seems like a bit of extra weight we don't need. We can just call the method directly.
```suggestion
            acts_dset.add_annotation(
                image_id=image_id,
                category_id=all_preds[i].item(),
                score=all_probs[i][all_preds[i]].item(),
                prob=all_probs[i].numpy().tolist(),
            )  
```

**Comment by cameron-a-johnson** [comment]
Changes accepted - thanks @Purg

**Review by Purg — APPROVED** [review]

**Pull request merged**